pkg/watcher/reconciler/dynamic/dynamic.go

```diff
@@ -388,19 +397,65 @@ func (r *Reconciler) streamLogs(ctx context.Context, o results.Object, logType,
 		return fmt.Errorf("error reading from tkn reader: %w", err)
 	}
 
-	errChanRepeater := make(chan error)
-	go func(echan <-chan error, o metav1.Object) {
-		writeErr := <-echan
-		errChanRepeater <- writeErr
+	errChanRepeater := make(chan error, 100) //some stuff on the internet says buffered channels are better for GC
 
-		_, err := writer.Flush()
+	// logctx is derived from ctx. Therefore, if ctx is cancelled (either explicitly through a call to its cancel
+	// function or when it reaches its deadline), logctx will be cancelled automatically.
+	logctx, _ := context.WithTimeout(ctx, 10*time.Minute)
+
+	go func(ctx context.Context, echan <-chan error, o metav1.Object) {
+		select {
+		case <-ctx.Done():
+			logger.Warnw("Context done streaming log",
+				zap.String("namespace", o.GetNamespace()),
+				zap.String("name", o.GetName()),
+			)
+		case writeErr := <-echan:
+			errChanRepeater <- writeErr
+			var gofuncerr error
+			_, gofuncerr = writer.Flush()
+			if gofuncerr != nil {
+				logger.Error(gofuncerr)
+			}
+			if gofuncerr = logsClient.CloseSend(); gofuncerr != nil {
```

Comment:
per comment from @adambkaplan in our internal discussion @ramessesii2 we want to execute the `CloseSend()` here regardless of which case in the select fires, meaning it needs to be moved out of this case, out of the select in fact, so that it is also executed when ctx.Done() is processed as well

Reply:
i.e release the streaming resources regardless if the tkn client library is not properly reconciling the error channel used to block the goroutine from completing 

that at least would help slow things .... cleaning up the threads (or moving to a different model) could become a separate concern based on what you find as you drill down on your suspicion

Reply:
So we've confirmed in an internal slack thread including @ramessesii2 @sayan-biswas and myself that the `streamLogs` code in the "main" thread needs to close the channel the goroutine is blocked on.

Reply:
OK @ramessesii2 your call to CloseSend minimally is still only in the `case writeErr := <-echan:` block

we need to call this as well if the `case <-ctx.Done():` block fire first ... i.e. there was some weird hang on the write, so we can abort and free up the resources

my initial assumption is the flush is irrelevant in this error case and case stay where it is, but do let us know if your testing uncovers a different conclusion

please update your PR so we can close this thread and have @pmacik 's next round of testing include this change

Reply:
done, I've done a bit more of refactoring. Hopefully, this is what you're asking as well here. Pushing the changes now.